### ISSUE TITLE
Dynamically loading assembly referencing Google.Cloud.Spanner.Data assembly results in binding redirect errors

### ISSUE THREAD

**Issue by FransBouma** [googleapis/google-cloud-dotnet#2953]
#### Environment details

  - OS: Windows 8.1/10
  - .NET version: 4.5.2
  - Package name and version: Google.Cloud.Spanner.Data v2.0.0 (beta8)

#### Steps to reproduce

  1. Create an assembly which references Google.Cloud.Spanner.Data v2.0.0 package from NuGet and opens a connection
  2. Dynamically load this assembly from a folder. This folder contains all assemblies (indirectly) referenced by Google.Cloud.Spanner.Data. The folder is added to the probing path of the application loading the assembly. (the application is a winforms application)

When doing so, you'll get various binding redirect errors and you have to add binding redirects to your application's app.config file. 

This is inconvenient and IMHO also unnecessary: the nuget package for Google.Cloud.Spanner.Data v2.0.0 references other nuget packages which result in having assemblies on disk with higher version numbers for some assemblies than what they directly reference. This requires binding redirects to be applied. For a project that directly references Google.Cloud.Spanner.Data, you could add a `AutoGenerateBindingRedirects` element to the csproj, but for a dynamically loaded assembly this isn't going to work. 

Will this be corrected in GA/RTM? If so, then ignore my report. If not, could it be made so that the assemblies on disk reference the assemblies that are also on disk so no binding redirects are required? 

TIA :)

**Comment by jskeet** [comment]
It's not clear what you mean by "corrected" here - what are you suggesting is *incorrect*, precisely?

There are several dependencies involved, and even if at the time we hit GA for Google.Cloud.Spanner.Data version 2.0 we make sure that every single dependency is at "latest", you'd get into exactly the same problem later on if you add a dependency to some other package which (say) depended on a later version of Google.Api.Gax.Grpc. So any "fix" could only have a very limited benefit.

Fundamentally, this is a platform problem - we follow SemVer, and expect the platform to support this, allowing us to have references to an earlier minor version of a package than some other dependency might have. I'm reluctant to go to huge lengths to apply a limited-effectiveness sticking plaster over a platform issue.

Of course, if I've misunderstood the problem or what you're suggesting, I'm very happy to be corrected :) I definitely want to make this work as reasonably as possible.

**Comment by jskeet** [comment]
(I've labelled this as a bug even though I don't strictly consider it to be one. It's as close as we've got, I think. Maybe feature request would be better...)

**Comment by FransBouma** [comment]
In my assembly, a driver for Google.Cloud.Spanner for LLBLGen Pro, I reference the nuget package, Google.Cloud.Spanner.Data v2, beta8. This brings some additional packages and all in all it results in additional files on disk. 

When I dynamically load my assembly, it by default doesn't deal with binding redirects. When I open a connection, I get: 

```
Message: The type initializer for 'Google.Cloud.Spanner.Data.SessionPoolManager' threw an exception.
Source: Google.Cloud.Spanner.Data
Stack trace: 
   at Google.Cloud.Spanner.Data.SessionPoolManager.get_Default()
   at Google.Cloud.Spanner.Data.SpannerConnection..ctor()
   at Google.Cloud.Spanner.Data.SpannerProviderFactory.CreateConnection()
   at SD.LLBLGen.Pro.DBDriverCore.DBDriverBase.CreateConnection() in C:\Myprojects\VS.NET Projects\LLBLGen Pro v5.5 Spanner\Designer\DBDriverCore\DBDriverBase.cs:line 548
   at SD.LLBLGen.Pro.Gui.Controls.WizardPages.MetaDataRetrievalWizard_Step_ConnectionData.TestConnectionData(Boolean showSuccess) in C:\Myprojects\VS.NET Projects\LLBLGen Pro v5.5 Spanner\Designer\Gui\Controls\WizardPages\MetaDataRetrievalWizard_Step_ConnectionData.cs:line 244

Inner exception:-----------------------
Exception details:
=====================
Message: Could not load file or assembly 'Google.Protobuf, Version=3.6.0.0, Culture=neutral, PublicKeyToken=a7d26565bac4d604' or one of its dependencies. The located assembly's manifest definition does not match the assembly reference. (Exception from HRESULT: 0x80131040)
Source: Google.Cloud.Spanner.Data
Stack trace: 
   at Google.Cloud.Spanner.Data.SessionPoolManager..cctor()
```
Here, the Google.Protobuf assembly v3.6.0.0 is referenced, however the file on disk is v3.6.1.0. This fails at runtime of course. 

Another one, when I resolved the one above:
```
Message: Could not load file or assembly 'Google.Apis.Auth, Version=1.21.0.0, Culture=neutral, PublicKeyToken=4b01fa6e34db77ab' or one of its dependencies. The located assembly's manifest definition does not match the assembly reference. (Exception from HRESULT: 0x80131040)
Source: Google.Cloud.Spanner.Data
Stack trace: 
   at Google.Cloud.Spanner.Data.SpannerClientCreationOptions.<GetCredentialsAsync>d__18.MoveNext()
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.Start[TStateMachine](TStateMachine& stateMachine)
   at Google.Cloud.Spanner.Data.SpannerClientCreationOptions.GetCredentialsAsync()
   at Google.Cloud.Spanner.Data.SessionPoolManager.<CreateClientAsync>d__26.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Google.Cloud.Spanner.Data.SessionPoolManager.TargetedPool.<>c__DisplayClass5_0.<<-ctor>g__CreateSessionPoolAsync|0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Google.Cloud.Spanner.Data.SpannerConnection.<>c__DisplayClass53_0.<<OpenAsyncImpl>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Google.Cloud.Spanner.Data.ExecuteHelper.<>c__DisplayClass1_0.<<WithErrorTranslationAndProfiling>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Google.Cloud.Spanner.Data.ExecuteHelper.<WithErrorTranslationAndProfiling>d__2`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Google.Api.Gax.TaskExtensions.WaitWithUnwrappedExceptions(Task task, TimeSpan timeout)
   at Google.Cloud.Spanner.Data.SpannerConnection.Open(Action transactionEnlister)
```

The other ones that weren't matching the versions on disk:

- Newtonsoft.Json (referenced 10.0.0, file on disk is 11.0.0)
- System.Interactive.Async (referenced 3.1.0.0, file on disk is 3.2.0.0)

The reason I asked was whether this was a result of the library being in beta, so things were still in flux, or whether this isn't fixable (as e.g. 2 projects reference the same package, 1 references version N, the other version N+1, and that one being the one on disk) or whether it's an oversight.

**Comment by jskeet** [comment]
I suspect the Google.Protobuf "issue" is because we depend (indirectly) on Google.Api.CommonProtos, which depends on Google.Protobuf 3.6.1, but we also depend (indirectly) on Grpc.Gcp depends on Google.Protobuf 3.6.0.

I don't know what depends on System.Interactive.Async 3.1.0, but I can see that Google.Api.Gax depends on 3.2.0.

I don't get to rerelease every dependency to make sure that everything in the entire tree lines up. In this case I "control" some of the packages, but not all of them - I'm not responsible for Grpc.Gcp, for example. Yes, I could get in touch with the people who release that, but if they update to 3.6.1 then it would be causing exactly the same problem for a different library which was using 3.6.0.

Even if I *could* do that, it's very unlikely that Spanner would be used entirely on its own - if it's in an ASP.NET Core app for example, that almost guarantees that the version of Newtonsoft.Json that ends up on disk is a later one than the earliest one referenced in the tree.

Basically, this is a bigger problem than any one package can solve - it's a platform issue. (And note that for the most part this is within a particular major version. We have a much bigger issue in terms of dependencies *across* major versions.)

I'm going to close this now as effectively infeasible. I'm sorry I don't have a happier answer for you :(

**Comment by FransBouma** [comment]
It's ok :) At least assembly redirects are possible and they solve it. I understand why it's not really solvable now and that's enough. :)
